Fix BBF PIVOT cannot use CTE as source table

RIC06X opened the pull request:
### Description

  Fixed BBF PIVOT stmt cannot use CTE as source table.
  Fixed string in pivot clumn value list caused crash
  Fixed conflicting column name is allowed in BBF PIVOT clause


### Issues Resolved

BABEL-4958, BABEL-4959, BABEL-4960, BABEL-4977

### Test Scenarios Covered ###
* **Use case based -**
    - string in PIVOT column value list (BABEL-4958)
    - PIVOT with CTE as source table
    - aggregate column in PIVOT column value list





### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

RIC06X committed:
Fix BBF PIVOT cannot use CTE as source table

Fixed BBF PIVOT stmt cannot use CTE as source table.
Fixed string in pivot clumn value list caused crash
Fixed conflicting column name is allowed in BBF PIVOT clause

Task: BABEL-4958, BABEL-4959, BABEL-4960
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

timchang514 commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y:
```diff
@@ -1640,33 +1643,34 @@ tsql_pivot_expr: TSQL_PIVOT '(' func_application FOR columnref IN_P in_expr ')'
-								subsel_valuelists = lappend(subsel_valuelists, l);
-							}
-						}
+						Node 	*conlumn_const;
```
should this be `column_const` instead?

RIC06X replied:
Thanks! I'll correct the typo

tanscorpio7 commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y:
```diff
@@ -1640,33 +1643,34 @@ tsql_pivot_expr: TSQL_PIVOT '(' func_application FOR columnref IN_P in_expr ')'
+
+					column_list = (List *)$8;
+					column_const_list = NIL;
+					value_col = list_nth_node(ColumnRef, ((FuncCall *) $3)->args, 0);
+					value_colstr = list_nth_node(String, value_col->fields, ((List *)value_col->fields)->length - 1);
```
looking at grammer for func_application, args could be NIL, should we add some NULL checks here ?

RIC06X replied:
Thanks for the review, I'll add a function args check in gram-tsql-rule.y

RIC06X replied:
Addressed this issue in grammar rule definition. 
In TSQL PIVOT definition, the aggregation function only support  1 column argument, so I changed the rule to the following one:
```
tsql_pivot_expr: TSQL_PIVOT '(' func_name '(' columnref ')' FOR columnref IN_P '(' columnList ')' ')'
```
This will restrict the aggregate function argument to be not null and must be a column ref and there could be only 1 argument in the aggregate function. We don't need to explicitly check it here because bison parser will check it for us.

xhfanhe commented on contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-rule.y:
```diff
@@ -1640,33 +1643,34 @@ tsql_pivot_expr: TSQL_PIVOT '(' func_application FOR columnref IN_P in_expr ')'
+						if (column_str != NULL && strcmp(column_str->sval, value_colstr->sval) == 0)
+							ereport(ERROR,
+									(errcode(ERRCODE_SYNTAX_ERROR),
+										errmsg("The column name \"%s\" specified in the PIVOT operator conflicts with the existing column name in the PIVOT argument.", value_colstr->sval),
```
Is this the same error msg as SQL Server?

RIC06X replied:
Yes, same error msg as SQL Server

tanscorpio7 commented:
Please add some test cases where source table is schema specified

tanscorpio7 commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -4971,6 +4978,7 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
 	/* use the copy of the orgininal fromClause to prevent double analyzing fromClause */
 	((SelectStmt *)stmt->srcSql)->fromClause = src_sql_fromClause_copy;
 	/* Transform the new src_sql & get the output type of that agg function*/
+	((SelectStmt *)stmt->srcSql)->withClause = with_clause;
```
If I understand correctly we are finally placing pivot_src_sql as the source sql in bbf_pivot_function.
could you point where pivot_src_sql-> fromClause & withClause go through transformation ?

RIC06X replied:
If the new created pivot_src_sql is a select stmt with pivot operator, then it will go through transform_pivot_clause hook again. If there is no more pivot in subquery,  pivot_src_sql is just a normal select stmt, and it will go through the original fromClause and withClause transformation.

forestkeeper commented on test/JDBC/expected/pivot-vu-verify.out:
```diff
@@ -1238,6 +1238,63 @@ Orange#!#17025.0000#!#8512.5000#!#77659.0000#!#38829.5000#!#4535.0000#!#23425.00
+    FROM orders [o] JOIN products AS [p] on (o.productId = p.productId)
+)
+SELECT * FROM cte_table
+PIVOT (
```
Let's try a test case like : 

With cte_table AS ()
select * from cte_table join (
      select * from another_table 
      PIVOT ...
) derived on derived.xx = cte_table.xx

Wanted to check different stmt level pivot and cte table are in.

RIC06X replied:
test cases added

RIC06X committed:
Fix table with schema name can't be recognized

Fixed table with schema name can't be recognized in pivot stmt

Task: BABEL-4977
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

tanscorpio7 commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -4968,9 +4991,15 @@ transform_pivot_clause(ParseState *pstate, SelectStmt *stmt)
+	((SelectStmt *)stmt->srcSql)->withClause = with_clause;
+
+	/* we rewrite srcSql object refereces again because we used a new copy of fromClause */
+	if (enable_schema_mapping())
+		rewrite_object_refs((Node *)stmt->srcSql);
```
Suggestion: lets add a test case where all objects are completely qualified.
for example babelfish allows db.schema.table.colname & db.schema.tablename

tanscorpio7 replied:
otherwise PR looks good
